docs/get_started/stable_c_abi.rst

```diff
@@ -0,0 +1,189 @@
+
+- ``handle: void*``: a library handle that contains the function to be called. It is usually ``nullptr`` for most use cases, but some libraries (such as cuDNN) may use it to store global contexts.
+- ``args: Any*``: a pointer to an array of ``Any`` values for input arguments, stored contiguously on stack or heap.
+- ``num_args: int``: number of input arguments.
+- ``result: Any*``: a pointer to a single ``Any`` that receives the output result.
+
+.. figure:: ../_static/tvm-ffi-layout-func.svg
+   :alt: Calling convention for universal_c_abi
+   :name: fig:layout-func
+
+   Figure 2. Layout spec of ``universal_c_abi`` function.
+
+
+Stability and Interoperability
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+**Stability.** The pure C layout and the calling convention are stable across compiler versions and independent of host languages or frameworks.
+
+**Cross-language.** TVM-FFI implements this calling convention in multiple languages (C, C++, Python, Rust, ...), enabling code written in one language—or generated by a DSL targeting the ABI—to be called from another language.
+
+**Cross-framework.** TVM-FFI uses standard data structures such as `DLPack tensors <https://data-apis.org/array-api/2024.12/design_topics/data_interchange.html#dlpack-an-in-memory-tensor-structure>`_ to represent arrays, so compiled functions can be used from any array framework that implements the DLPack protocol (NumPy, PyTorch, TensorFlow, CuPy, JAX, and others).
+
+
+Stable ABI in C Code
+--------------------
+
+.. note::
+
+  All the code in this tutorial can be found under `examples/stable_c_abi <https://github.com/apache/tvm-ffi/tree/main/examples/stable_c_abi>`_ in the repository, including commands to compile the code and run the examples.
+
+This section shows how to write C code that follows the stable C ABI. Specifically, we provide two examples:
+
+- A CPU ``add_one_cpu`` kernel in C that is equivalent to the :ref:`C++ example <cpp_add_one_kernel>`.
+- A loader and runner in C that invokes the kernel, a direct C translation of the :ref:`C++ example <cpp_load>`.
+
+The first shows **how to write a function** that follows the stable C ABI; the second shows **how to call a function** via the stable C ABI.
+
+C Kernel ``add_one_cpu``
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Below is a minimal ``add_one_cpu`` kernel in C that follows the stable C ABI. It has three steps:
+
+- **Step 1**. Extract input ``x`` and output ``y`` as DLPack tensors;
+- **Step 2**. Implement the kernel ``y = x + 1`` on CPU with a simple for-loop;
+- **Step 3**. Set the output result to ``result``.
+
+.. literalinclude:: ../../examples/stable_c_abi/src/add_one_cpu.c
+   :language: c
+   :start-after: [example.begin]
+   :end-before: [example.end]
+
+
+Compared to the :ref:`C++ example <cpp_add_one_kernel>`, there are a few key differences:
+
+- The explicit marshalling in **Step 1** is only needed in C. In C++, templates hide these details.
```

Comment:
it is better to directly talk about this example, target audience here is compiler developers, we can add c++ template reference in end.

At high-level, we perform the following steps:

- Function symbols follow name int __tvm_ffi_<name>
- We go over and check the type_index in  TVMFFIAny and extract the input function arguments
- The function Return 0 for success, -1 for error 
    - When error happens, we set the error via `TVMFFIErrorSetRaisedFromCStr` or `TVMFFIErrorSetRaisedFromCStrParts`

Reply:
adjusted a little bit:

> **Compiler codegen**. This C code serves as a direct reference for DSL compilers. To emit a function that follows the stable C ABI, the compiler needs to ensure the following aspects:
> 
> - Symbol naming: define the exported symbol name as ``__tvm_ffi_{func_name}``;
> - Type checking: check types of input arguments via :cpp:member:`TVMFFIAny::type_index`, and then marshal inputs from :cpp:class:`TVMFFIAny` to the desired types;
> - Exception handling: Return 0 on success, or a non-zero error code on failure. When error occurs, set appropriate error message via :cpp:func:`TVMFFIErrorSetRaisedFromCStr` or :cpp:func:`TVMFFIErrorSetRaisedFromCStrParts`.